### PR TITLE
Improve ManuallyDrop suggestion

### LINKED ISSUE #90585: Suggestion for E0740 (union contains a field that needs dropping) doesn't change anything
Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dd41f5066ca17cbf7f00f6f9349b9c1d)):

```rust
union U {
    value: Vec<isize>,
}
```

The current output is:

```
error[E0740]: unions may not contain fields that need dropping
 --> src/lib.rs:2:5
  |
2 |     value: Vec<isize>,
  |     ^^^^^^^^^^^^^^^^^
  |
note: `std::mem::ManuallyDrop` can be used to wrap the type
 --> src/lib.rs:2:5
  |
2 |     value: Vec<isize>,
  |     ^^^^^^^^^^^^^^^^^
```

Ideally the output should look like:

```
error[E0740]: unions may not contain fields that need dropping
 --> src/lib.rs:2:5
  |
2 |     value: Vec<isize>,
  |     ^^^^^^^^^^^^^^^^^
  |
note: `std::mem::ManuallyDrop` can be used to wrap the type
 --> src/lib.rs:2:5
  |
2 |     value: std::mem::ManuallyDrop<Vec<isize>>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by rukai**
closes https://github.com/rust-lang/rust/issues/90585
* Fixes the recommended change to use ManuallyDrop as per the issue
* Changes the note to a help
* improves the span so it only points at the type.

### CODE REVIEW

File: compiler/rustc_typeck/src/check/check.rs [hunk 1]
```diff
@@ -371,16 +371,25 @@ fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) -> b
...
             let field_ty = field.ty(tcx, substs);
-            // We are currently checking the type this field came from, so it must be local.
-            let field_span = tcx.hir().span_if_local(field.did).unwrap();
+            let (field_span, ty_span) =
+                // We are currently checking the type this field came from, so it must be local.
+                if let Node::Field(field) = tcx.hir().get_if_local(field.did).unwrap() {
+                    (field.span, field.ty.span)
+                } else {
+                    unreachable!("mir field has to correspond to hir field");
+                };
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/check.rs:380]:
```suggestion
            let (field_span, ty_span) = match tcx.hir().get_if_local(field.did) {
                // We are currently checking the type this field came from, so it must be local.
                Some(Node::Field(field)) => (field.span, field.ty.span),
                _ => unreachable!("mir field has to correspond to hir field"),
            };
```

File: compiler/rustc_typeck/src/check/check.rs [hunk 2]
```diff
@@ -371,16 +371,25 @@ fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) -> b
...
                 struct_span_err!(
                     tcx.sess,
                     field_span,
                     E0740,
                     "unions may not contain fields that need dropping"
                 )
-                .span_note(field_span, "`std::mem::ManuallyDrop` can be used to wrap the type")
+                .multipart_suggestion_verbose(
+                    "wrap the type with `std::mem::ManuallyDrop` and ensure it is manually dropped",
+                    vec![(ty_span, format!("std::mem::ManuallyDrop<{}>", field_ty))],
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/check.rs:390]:
```suggestion
                    vec![
                        (ty_span.shrink_to_lo(), "std::mem::ManuallyDrop<".to_string()),
                        (ty_span.shrink_to_hi(), ">".to_string()),
                    ],
```

Comment (rukai) [reply compiler/rustc_typeck/src/check/check.rs:390]: oh thats clever!

**Comment by rukai** [comment]
This is ready for rereview

**Pull request merged**